### ISSUE TITLE
INI parser overwrites values

### ISSUE THREAD

**Issue by basilfx** [knxd/knxd#315]
With 0.14.17 and `master`, I have noticed (using GDB) that the INI parser overwrites values, therefore forking and never printing to stdout (even with `background = false`). I have a similar result on a different machine (Ubuntu).

Steps:
* Compile master without systemd.
* Start GDB using `gdb --args src/server/knxd knx.ini`, where `knx.ini` contains the config file below.
* Step until line 342, observe that `logfile` and `pidfile` are empty.
* Step until line 346, observe that `logfile` and `pidfile` are (suddenly) equal to `background`.
* Observe that the expression on line 356 will now evaluate to true, even though `logfile` was empty.

Screenshot:
![schermafbeelding 2018-01-18 om 23 50 37](https://user-images.githubusercontent.com/815976/35125609-90496206-fcaa-11e7-914b-47685e71e685.png)

Config file: 
```
[A.tcp]
debug = debug
server = knxd_tcp
systemd-ignore = true

[B.dummy]
debug = debug
driver = dummy

[debug]
error-level = 0xffffffff
trace-mask = 0xffffffff

[main]
addr = 0.0.1
client-addrs = 0.0.20:9
connections = A.tcp,B.dummy
debug = debug
```

**Comment by tru7** [comment]
Just wanted to open my own issue. Same problem here: no output for -t when knxd is started in shell. This worked until commit 6b5a683d19a3d82488faa41b29e04bfcf874485a

**Comment by tru7** [comment]
I think logic in line 364 is invalid
`if (logfile && *logfile || !num_fds)`
this means stderr and stdout are pointed to /dev/null if knxd is run from shell in foreground without naming a logfile.
Maybe it's OK to point stderr and stdout to /dev/null if knxd is run in background like this
`if (logfile && *logfile || background)`

**Comment by tru7** [comment]
I guess this is sufficient for line 364
`if (background)`
resulting stderr and stdout pointing to logfile if there is one or /dev/null otherwise only if knxd is running in background. If knxd is running in foreground or controlled by systemd there is no redirect for stdout and stderr.

**Comment by smurfix** [comment]
Actually it's not. You can run in the foreground with a logfile. That makes sense when debugging.

I'm preparing 0.14.23 which fixes this correctly.

**Comment by basilfx** [comment]
This doesn't fix the original issue. I just recompiled `master`, and I still do not get any output.

Using GDB, I can still confirm that the `logfile` and `pidfile` are initially `''`, until `background = main->value("background",false);` is evaluated (actually, it already changes when it steps into the `value()` function.

Any idea's?

**Comment by smurfix** [comment]
Please compile with -O0 (no optimization) before deciding which values get changed where.

**Comment by smurfix** [comment]
I can't reproduce that.
```
src/server/knxd -t1023 -e1.2.3 -E1.2.4:5 -b ip: -b dummy:
Layer 4 [ 1:main        0.000] initialized
Layer 4 [ 1:main        0.000] setting up
…
```

**Comment by basilfx** [comment]
I gave it a shot using GCC-5 instead of Clang-9, and that makes a difference:

```
basilfx:knxd/ (master✗) $ src/server/knxd -t1023 -e1.2.3 -E1.2.4:5 -b ip: -b dummy:
knxd: invalid option -- t
Layer 4 [ 1:main        0.000] initialized
Layer 4 [ 1:main        0.000] setting up
E00000055: [ 1:main] Driver 'ip' not found.
E00000055: [ 1:main] Section 'A.ip' has no known server nor driver.
Layer 4 [ 1:main        0.000] setup BROKEN
F00000000: [ 1:main] Error setting up the KNX router.
```

Since I had a similar result (but different reason) under Linux, my first thought wasn't the compiler. So, I guess I can close this issue now.

Thanks for the help and sorry for the inconvenience.

**Comment by smurfix** [comment]
Umm, how the ??? is `-t` an invalid option?

**Comment by basilfx** [comment]
Good question. If I put something else before `-t`, that other option becomes invalid. Guess the macOS build is more broken than expected.

**Comment by tru7** [comment]
I have the same issue when crosscompiling for LEDE / OpenWRT as mentioned in #266. Every first argument is considered invalid. It only works when starting with ini file.

**Comment by smurfix** [comment]
Well, that does look like a broken argp implementation. Please try downloading the current version and compiling knxd_args with that.

**Comment by tru7** [comment]
I doubt that, argp-standalone is the latest stable 1.3 and has ever been, also when it still worked with knxd.
Crosscompiled knxd_args does not have a problem with the parameters, only knxd has.

**Comment by smurfix** [comment]
Strange. knxd sets the ARGP_NO_ERRS flag which is supposed to not print any errors. It then passed the whole argument list off to knxd_args. Can you `strace -f` this to check whether the argument list gets mangled?

**Comment by tru7** [comment]
Unfortunately debugging is hard on my LEDE device. There is no strace command and cannot be installed. Maybe this got a little off-topic from the original post. Should I open a different issue to start over?
I found some more problems with knxd 0.14.23 on LEDE: I cannot get it to log into logfile, it's opened but never written to. I also cannot get it to write to stderr when using the ini file without logfile and running in foreground.
BTW: knxd -V works, but not with any other option.
The only way to make knxd run is to first create a knxd.ini file with knxd_args and then starting knxd with the ini file.

**Comment by tru7** [comment]
correction: I needed to update the package list after reboot. Now I have strace installed.
Strange result: it works fine when using strace to start knxd with arguments. foreground and background

**Comment by smurfix** [comment]
That smells like an uninitialized variable.

**Comment by tru7** [comment]
some more corrections, sorry about that.
Somehow I am not free in selecting a path for logfile on this platform. When I use /tmp/knxd.log it works but does not when using /var/log/knxd.log. I don't understand this (yet).

Now I see log output in /tmp/knxd.log. And with the 0.14.23 version I also see output in stderr when running in foreground. My former results were probably from older test with a commit when stderr was bound to /dev/null.
I was also confused when starting knxd with arguments. I just did not realise that knxd started up fine the first time even when complaining about invalid argument. Of course it could not start on further attemps because it was already running. But without output in stderr I did not realize.
To make it short - knxd starts up fine now with arguments but it complains about invalid argument for the first one. With strace I just missed this line:
knxd: unrecognized option `--eibaddr=0.0.20'

**Comment by smurfix** [comment]
> does not when using /var/log/knxd.log

What are the rights to that file, or the directory?

> knxd: unrecognized option `--eibaddr=0.0.20'

Is that the first option? if so, it's the same problem as above, i.e. looks like your argp ignores the ARGP_NO_ERRS flag.

**Comment by tru7** [comment]
Yes, it's the first option, but it's still used for the knxd process. I don't know about ARGP_NO_ERRS. argp-standalone is maintained in the LEDE/OpenWrt firmware repo at https://github.com/openwrt/openwrt/tree/master/package/libs/argp-standalone

With strace I found /tmp/knxd.log is correctly opened but /var/log/knxd.log is opened as "\267\236\343\230\267\236\343\230er-setup" and is found in the local directory. So back to the original poster's problem?

**Comment by tru7** [comment]
root@LEDE:~# ls -ld /tmp /var/log
drwxrwxrwt   19 root     root           500 Jan 21 12:09 /tmp
drwxrwxr-x    2 root     root            80 Dec 22 17:30 /var/log

**Comment by smurfix** [comment]
Right. So you don't have rights for /var/log, and the problem is that there's no error message.

**Comment by smurfix** [comment]
… or something. It seems that I can't depend on argp not modifying the vector when I fork off during argp processing. The fix shall be to check that argc==2 (or 3 (name of main section)) and argv[1][0] != '-' (it's a config file!) – everything else shall get punted to the knxd_args helper.

You want to fix it yourself and send a pull request? it'll be a few days for me to get around to that.

**Comment by tru7** [comment]
I'm not a developper so I won't be able to do that but I'm patient and appreciate your time.
I don't think it's a problem of rights because everything is run as root, it's a router firmware after all.
root@LEDE:~# ls -l /tmp/knxd.log
-rw-r-----    1 root     root         17108 Jan 21 13:04 /tmp/knxd.log

in the meantime I tried to replace the default -Os by -O0 but there is no difference. The logfile name is parsed correctly but where it's opened it's some invalid, unusable name when it fails.

it looks like it depends on the name length. /tmp/knxd12.log and /var/log/knxd12 work but /tmp/knxd123.log does not

**Comment by tru7** [comment]
I found some bug: assigning logfile and pidfile in knxd.cpp lines 337 and 341 is somehow erroneous because it points to nonsens if length of filename (including NULL) is greater than 16. However main->value pointer shows correct string value when I use a printf for it.
This happens for:
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.5)
gcc version 6.3.0 20170516 (Debian 6.3.0-18)
No problem for:
gcc version 4.9.2 (Debian 4.9.2-10)

**Comment by tru7** [comment]
I do not know coding but I found this for c_str() in the manual "A program shall not alter any of the characters in this sequence." and "The pointer returned may be invalidated by further calls to other member functions that modify the object." Could that mean it's invalid to assign values to logfile and pidfile pointers?

**Comment by smurfix** [comment]
Thanks for spotting this. I'll prepare an update soon(ish).

**Comment by Race666** [comment]
@basilfx  Would like to try the patch from #326 if this solves your issue?

**Comment by tru7** [comment]
@smurfix I think with commit https://github.com/knxd/knxd/commit/d29f7047a8b20a7ac57c3c9ad349c12be60f9fee the problems with filenames for logfile and pid should be fixed.
As for invalid (first) options I have spent some time looking into arp-standalone. I come to the conclusion that it's a bug in the source code of argp-standalone by not honouring ARGP_NO_ERRS for at least "illegal option" and "invalid option". This would be quite easy to fix but nobody cared since 2004-02-23 and the project looks unmaintained since about then. I have set up a demo for it at https://github.com/tru7/argp-demo which should show the bug on any platform not using (g)libc.
Maybe this issue can be closed?
